### PR TITLE
[Clustering] Average Gradient Aggregation

### PR DESCRIPTION
**Pull request by johan-gras**
This PR is two fold: a new feature Average Gradient Aggregation and a cleaning of the clustering codebase.

## Average Gradient Aggregation
Average Gradient Aggregation feature averages the gradient of each cluster centroid by the size of this cluster (the size being the number of child's weight belonging to the respective cluster).
### User API
We are letting users the choice between the existing implementation (Sum Gradient Aggregation) and the new implementation (Average Gradient Aggregation):
```python
tfmot.clustering.keras.cluster_weights(
    to_cluster, number_of_clusters, cluster_centroids_init, cluster_gradient_aggregation=GradientAggregation.SUM
)
```
With cluster_gradient_aggregation being an enum: either `GradientAggregation.SUM` (default) or `GradientAggregation.AVG`.
#### Sum Gradient Aggregation
sum_grad
With Sum Gradient Aggregation, the gradient of each centroid is simply computed through normal backpropagation: the gather operation will aggregate the gradient of the corresponding weights child's to their centroids. **This is a summation of gradients.**
`clustered_weight = tf.gather(cluster_centroids, pulling_indices)`
#### Average Gradient Aggregation
avg_grad
With Average Gradient Aggregation, the gradient of each centroids is modified with a `tf.custom_gradient` by dividing the current gradient by the size of the cluster (number of belonging weight's child). **This is an averaged summation of gradients.**
```python
@tf.custom_gradient
def average_centroids_gradient_by_cluster_size(cluster_centroids, cluster_sizes):
  def grad(d_cluster_centroids):
    # Average the gradient based on the number of weights belonging to each cluster
    d_cluster_centroids = tf.math.divide_no_nan(d_cluster_centroids, cluster_sizes)
    return d_cluster_centroids, None
  return cluster_centroids, grad

# Compute the size of each cluster (number of weights belonging to each cluster)
cluster_sizes = tf.math.bincount(
  arr=tf.cast(pulling_indices, dtype=tf.int32),
  minlength=tf.size(cluster_centroids),
  dtype=cluster_centroids.dtype,
)
# Modify the gradient of cluster_centroids to be averaged by cluster sizes
cluster_centroids = average_centroids_gradient_by_cluster_size(
  cluster_centroids,
  tf.stop_gradient(cluster_sizes),
)
# Gather the clustered weights based on cluster centroids and pulling indices
clustered_weight = tf.gather(cluster_centroids, pulling_indices)
```
### Results
Average Gradient Aggregation doesn't show strict performance improvement across all possible baselines, but does show improvement over some baselines. Based on these results, we decided to let users decide between the two aggregation methods, with Sum Gradient Aggregation being the default setting for backward-compatibility.

| Model      | Dataset | Number of clusters | SumGA accuracy | AvgGA accuracy | Delta accuracy |
| ----------- | ----------- | ----------- | ----------- | ----------- | ----------- |
| DS-CNN-L | Speech Commands | 8 | 94.03%  | 94.11% | +0.08% |
| MobileNetV2 | Imagenet  | 16 | 67.76% | 68.39% | +0.63% |
| DS-CNN-L | Speech Commands | 32 | 95.17%  | 95.25% | +0.08% |
| MobileNetV2 | Imagenet  | 32 | 69.658% | 69.664% | +0.006% |
| MobileNetV1 | Cifar10 | 16 | 94.40% | 94.43% | +0.03% |

## Clustering codebase cleaning
This cleaning was the first step allowing the implementation of the Average Gradient Aggregation feature.

- Fix bugs with all possible regularizers (kernel, bias and activity).
- clustering_algorithm.py: removed non-necessary reshape operations and fuse everything into get_clustered_weight.
- cluster_wrapper.py: simplification of the logic in the build and call function.
- cluster.py: simplification of the strip_clustering function.
- Addition of a series of tests.

### CODE REVIEW

**Review by wwwind — APPROVED** [review]

**Review by daverim — APPROVED** [review]
Whoops sorry akarmi@ i rerequested your review accidentally.

**Review by wwwind — APPROVED** [review]

**Comment by johan-gras** [comment]
Thanks @daverim! Also, is there a problem with CI on this PR "Google internal checks FAILED..."?

**Review by akarmi — APPROVED** [review]

**Comment by daverim** [comment]
Hi there was, but I fixed it on our side.
